### PR TITLE
[AutoDiff] Handle `@inout_aliasable` when converting associated functions

### PR DESCRIPTION
**Pull request by rxwei**
When reapplying a `partial_apply` for a JVP/VJP, if the argument has convention `@inout_aliasable`, it is a noescape mutable capture and the underlying value should not be retained. This fixes a memory leak when a noescape closure being differentiated captures a mutable self.

Resolves https://github.com/eaplatanios/swift-ale/issues/1.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: lib/SILOptimizer/Mandatory/Differentiation.cpp [hunk 1]
```diff
@@ -1951,19 +1951,35 @@ reapplyFunctionConversion(SILValue newFunc, SILValue oldFunc,
         builder.createDeallocStack(loc, alloc);
     };
     for (auto arg : pai->getArguments()) {
-      // Retain the argument since it's to be owned by the newly created
+      // Retain the argument if it's to be owned by the newly created
       // closure.
+      // Objects are to be retained.
       if (arg->getType().isObject()) {
         builder.createRetainValue(loc, arg, builder.getDefaultAtomicity());
         newArgs.push_back(arg);
```
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Minor: how about using early `continue` after each call to `newArgs.push_back` to reduce indentation?

**Review by dan-zheng — APPROVED** [review]
Nice! I wonder if there are other argument conventions that require special handling?

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done. PTAL.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
> Nice! I wonder if there are other argument conventions that require special handling?

It appears that this is the only special case. I asked [a question](https://forums.swift.org/t/does-partial-apply-not-take-ownership-of-inout-aliasable-arguments/27642) on the forum.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp (hunk 1 above)
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Looks great!

**Review by dan-zheng — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**